hardware/coinft/src/CoinFT.cpp

```diff
@@ -0,0 +1,250 @@
+    for (size_t i = 0; i < values.size(); ++i) {
+        mat(i / cols, i % cols) = values[i];
+    }
+
+    return mat;
+}
+
+CoinFT::CoinFT(const std::string& port, unsigned int baud_rate, const std::string& calibration_file)
+    : io(), serial(io, port), running(false), reading_counter(0), tareFlag(false), tareSampleCount(0), tareSampleTarget(100) {
+    serial.set_option(boost::asio::serial_port_base::baud_rate(baud_rate));
+    std::cout << "Connected to Comport" << std::endl;
+    initializeSensor();
+
+    // Load the calibration matrix
+    try {
+        calibrationMatrix = readMatrixFromCSV(calibration_file);
+        std::cout << "Calibration matrix loaded. Dimensions: " << calibrationMatrix.rows() << " x " << calibrationMatrix.cols() << std::endl;
+
+        // Check that calibration matrix is 6x24
+        if (calibrationMatrix.rows() != 6 || calibrationMatrix.cols() != 24) {
+            std::cerr << "Error: Calibration matrix must be 6x24." << std::endl;
+        }
+    } catch (const std::exception& e) {
+        std::cerr << "Error loading calibration matrix: " << e.what() << std::endl;
+    }
+
+    // Initialize tareOffset to zero
+    tareOffset = Eigen::VectorXd::Zero(num_sensors);
+
+    // Initialize tareSamples
+    tareSamples.clear();
+}
+
+CoinFT::~CoinFT() {
+    stopStreaming();
+    closePort();
+}
+
+void CoinFT::initializeSensor() {
+    // Send IDLE command
+    sendChar(IDLE);
+    std::this_thread::sleep_for(std::chrono::milliseconds(10));
+
+    // Send QUERY command to get packet size
+    sendChar(QUERY);
+    std::this_thread::sleep_for(std::chrono::milliseconds(10));
+
+    // Read packet size
+    std::vector<uint8_t> data = readData(1);
+    if (!data.empty()) {
+        packet_size = static_cast<int>(data[0]) - 1;
+    } else {
+        throw std::runtime_error("Failed to read packet size.");
+    }
+
+    num_sensors = (packet_size - 1) / 2;
+    std::cout << "Number of sensors: " << num_sensors << std::endl;
+}
+
+void CoinFT::sendChar(uint8_t cmd) {
+    boost::asio::write(serial, boost::asio::buffer(&cmd, 1));
+}
+
+std::vector<uint8_t> CoinFT::readData(size_t length) {
+    std::vector<uint8_t> buffer(length);
+    boost::asio::read(serial, boost::asio::buffer(buffer));
+    return buffer;
+}
+
+void CoinFT::startStreaming() {
+    sendChar(STREAM);
+    running = true;
+    tareFlag.store(true); // Set tareFlag to true for initial taring
+    data_thread = std::thread(&CoinFT::dataAcquisitionLoop, this);
+}
+
+void CoinFT::stopStreaming() {
+    sendChar(IDLE);
+    if (running) {
+        running = false;
+        if (data_thread.joinable()) {
+            data_thread.join();
+        }
+    }
+    tareFlag.store(true); // Reset tareFlag so taring occurs next time
+}
+
+void CoinFT::tare() {
+    tareFlag.store(true);
+}
+
+void CoinFT::dataAcquisitionLoop() {
+    while (running) {
+        try {
+            // Read raw data
+            std::vector<uint16_t> processedPacket = readRawData();
```

Comment:
In a high rate real time loop, try to avoid memory allocations. All data variables such as `processedPacket`, `rawInputs` should be created and initialized outside of the loop.